File: sdk/python/src/plinth/workflows.py
# SPDX-License-Identifier: Apache-2.0
# Copyright 2026 The Plinth Authors
"""Workspace workflows: durable, resumable agent pipelines.

A workflow is a manifest of expected step names plus a server-tracked
log of completed steps. Each step has a lifecycle of
``pending -> running -> (completed | failed | cancelled)`` and may
reference a workspace snapshot at completion time so a crashed agent
can resume from a known checkpoint.

Two public surface elements:

* :class:`WorkflowsProxy` -- reachable via ``ws.workflows``. Owns
  ``create`` / ``get_or_create`` / ``get`` / ``list`` and returns
  :class:`WorkflowHandle` objects (never the bare model) so callers can
  chain step transitions ergonomically.
* :class:`WorkflowHandle` -- a thin wrapper around a :class:`Workflow`
  that exposes method-style ``start_step`` / ``complete_step`` / ...
  Holds a reference to the parent :class:`Workspace` so callers don't
  have to thread it through.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from .exceptions import (
    InvalidWorkflowStep,
    WorkflowNotFound,
    WorkflowStepNotFound,
    WorkspaceNotFound,
)
from .models import DLQEntry, Lease, ResumeInfo, Workflow, WorkflowStep

if TYPE_CHECKING:
    from .workspace import Workspace


# ---------------------------------------------------------------------------
# WorkflowHandle
# ---------------------------------------------------------------------------


class WorkflowHandle:
    """Method-style wrapper around a :class:`Workflow`.

    Returned by :meth:`WorkflowsProxy.create`, :meth:`WorkflowsProxy.get`,
    and :meth:`WorkflowsProxy.get_or_create`. Holds a reference to the
    parent :class:`Workspace` so it can issue API calls without forcing
    the caller to pass the workspace through every step transition.

    Mutations (start_step, complete_step, ...) update the cached
    :attr:`model` so subsequent reads against the handle are consistent.
    """

    def __init__(self, workspace: Workspace, model: Workflow) -> None:
        self._ws = workspace
        self._wf = model
        # v1.1 — per-step retry config supplied at workflow create time.
        # Populated by :meth:`WorkflowsProxy.create` when the caller
        # passes a list of dicts; left empty for the v1.0 list[str]
        # path. Looked up by step name in :meth:`start_step` so the
        # caller doesn't have to repeat the config on every start.
        self._retry_config: dict[str, dict[str, Any]] = {}

    # -- attributes ----------------------------------------------------

    @property
    def id(self) -> str:
        """The workflow ID (``wf_<ulid>``)."""
        return self._wf.id

    @property
    def name(self) -> str:
        """The workflow name."""
        return self._wf.name

    @property
    def status(self) -> str:
        """Current workflow status from the cached model.

        This reflects the server's view as of the last refresh -- call
        :meth:`refresh` if you need to re-read state after an external
        change.
        """
        return self._wf.status

    @property
    def steps(self) -> list[WorkflowStep]:
        """The cached list of recorded :class:`WorkflowStep` entries."""
        return self._wf.steps

    @property
    def steps_manifest(self) -> list[str]:
        """The expected step names in declaration order."""
        return self._wf.steps_manifest

    @property
    def metadata(self) -> dict[str, Any]:
        """The workflow's free-form metadata dict."""
        return self._wf.metadata

    @property
    def model(self) -> Workflow:
        """The underlying :class:`Workflow` Pydantic model (cached)."""
        return self._wf

    def __repr__(self) -> str:  # pragma: no cover - cosmetic
        return (
            f"WorkflowHandle(id={self.id!r}, name={self.name!r}, "
            f"status={self.status!r})"
        )

    # -- step transitions ----------------------------------------------

    def start_step(
        self,
        name: str,
        input: Any = None,  # noqa: A002 - mirrors API field name
        *,
        snapshot_id: str | None = None,
        initial_status: str = "running",
        max_attempts: int = 1,
        retry_policy: str = "none",
        retry_initial_delay_seconds: float = 1.0,
        retry_max_delay_seconds: float = 60.0,
        retry_jitter: bool = True,
    ) -> WorkflowStep:
        """Create a step on this workflow.

        ``initial_status`` defaults to ``"running"`` (the v0.2 in-process
        flow where the agent starts work immediately). Pass
        ``initial_status="pending"`` for the durable workflow executor:
        the step will be visible to workers via ``pending_steps()`` and
        will only flip to ``running`` when a worker leases it.

        A subsequent :meth:`complete_step` / :meth:`fail_step` /
        :meth:`cancel_step` advances the step to a terminal state.

        v1.1: ``max_attempts`` / ``retry_policy`` /
        ``retry_initial_delay_seconds`` / ``retry_max_delay_seconds`` /
        ``retry_jitter`` configure the per-step retry policy. Defaults
        preserve v1.0 behaviour (``max_attempts=1``).

        Args:
            name: Must be one of the names in :attr:`steps_manifest`.
            input: Optional input payload to record on the step.
            snapshot_id: Optional snapshot taken before the step ran.
            initial_status: ``"running"`` (default) | ``"pending"``.
            max_attempts: Total attempts before the step lands in the DLQ.
            retry_policy: ``"none"`` | ``"exponential"`` | ``"fixed"``.
            retry_initial_delay_seconds: First retry delay base.
            retry_max_delay_seconds: Cap applied after exponential growth.
            retry_jitter: When True, ±25% jitter is applied to each delay.

        Raises:
            InvalidWorkflowStep: When ``name`` is not part of the manifest.
        """
        if self._wf.steps_manifest and name not in self._wf.steps_manifest:
            raise InvalidWorkflowStep(
                f"Step {name!r} is not declared in the workflow manifest "
                f"{list(self._wf.steps_manifest)!r}.",
            )

        # If the workflow was created with per-step retry config, fall
        # back to that config when the caller didn't override it. The
        # explicit kwargs always win over the cached config.
        cached = self._retry_config.get(name, {})
        if max_attempts == 1 and "max_attempts" in cached:
            max_attempts = int(cached["max_attempts"])
        if retry_policy == "none" and "retry_policy" in cached:
            retry_policy = str(cached["retry_policy"])
        if (
            retry_initial_delay_seconds == 1.0
            and "retry_initial_delay_seconds" in cached
        ):
            retry_initial_delay_seconds = float(
                cached["retry_initial_delay_seconds"]
            )
        if (
            retry_max_delay_seconds == 60.0
            and "retry_max_delay_seconds" in cached
        ):
            retry_max_delay_seconds = float(cached["retry_max_delay_seconds"])
        if retry_jitter is True and "retry_jitter" in cached:
            retry_jitter = bool(cached["retry_jitter"])

        body: dict[str, Any] = {"name": name, "initial_status": initial_status}
        if input is not None:
            body["input"] = input
        if snapshot_id is not None:
            body["snapshot_id"] = snapshot_id
        # v1.1 retry params — only sent when not the v1.0 defaults so the
        # request body stays small for callers that haven't opted in.
        if max_attempts != 1:
            body["max_attempts"] = max_attempts
        if retry_policy != "none":
            body["retry_policy"] = retry_policy
        if retry_initial_delay_seconds != 1.0:
            body["retry_initial_delay_seconds"] = retry_initial_delay_seconds
        if retry_max_delay_seconds != 60.0:
            body["retry_max_delay_seconds"] = retry_max_delay_seconds
        if retry_jitter is not True:
            body["retry_jitter"] = retry_jitter

        response = self._ws._http.post(
            f"/v1/workspaces/{self._ws.id}/workflows/{self.id}/steps",
            json=body,
            not_found_class=WorkflowNotFound,
        )
        step = WorkflowStep.model_validate(response.json())
        self._record_step(step)
        return step

    def complete_step(
        self,
        step_id: str,
        output: Any = None,
        *,
        snapshot_id: str | None = None,
    ) -> WorkflowStep:
        """Mark ``step_id`` completed, recording its output and snapshot.

        ``snapshot_id`` is the canonical resume point -- :meth:`resume_info`
        will surface it to the next agent.
        """
        return self._patch_step(
            step_id,
            status="completed",
            output=output,
            snapshot_id=snapshot_id,
        )

    def fail_step(
        self,
        step_id: str,
        error: str,
        *,
        output: Any = None,
    ) -> WorkflowStep:
        """Mark ``step_id`` failed with ``error``."""
        return self._patch_step(
            step_id,
            status="failed",
            error=error,
            output=output,
        )

    def cancel_step(self, step_id: str) -> WorkflowStep:
        """Mark ``step_id`` cancelled."""
        return self._patch_step(step_id, status="cancelled")

    def _patch_step(
        self,
        step_id: str,
        *,
        status: str,
        output: Any = None,
        error: str | None = None,
        snapshot_id: str | None = None,
    ) -> WorkflowStep:
        """Internal: PATCH a step into a new state."""
        body: dict[str, Any] = {"status": status}
        if output is not None:
            body["output"] = output
        if error is not None:
            body["error"] = error
        if snapshot_id is not None:
            body["snapshot_id"] = snapshot_id

        # The HTTP wrapper does not expose a verb-specific PATCH helper,
        # so fall through to the underlying httpx client and reuse its
        # error-mapping path.
        response = self._ws._http._client.request(
            "PATCH",
            f"/v1/workspaces/{self._ws.id}/workflows/{self.id}/steps/{step_id}",
            json=body,
        )
        self._ws._http._raise_for_status(
            response,
            not_found_class=WorkflowStepNotFound,
        )
        step = WorkflowStep.model_validate(response.json())
        self._record_step(step)
        return step

    # -- whole-workflow operations -------------------------------------

    def cancel(self) -> None:
        """Cancel the entire workflow.

        Refreshes the cached :class:`Workflow` so :attr:`status` reflects
        the new server state.
        """
        response = self._ws._http.post(
            f"/v1/workspaces/{self._ws.id}/workflows/{self.id}/cancel",
            not_found_class=WorkflowNotFound,
        )
        self._wf = Workflow.model_validate(response.json())

    def resume_info(self) -> ResumeInfo:
        """Return the next step + snapshot to resume from after a crash."""
        data = self._ws._http.get_json(
            f"/v1/workspaces/{self._ws.id}/workflows/{self.id}/resume",
            not_found_class=WorkflowNotFound,
        )
        return ResumeInfo.model_validate(data)

    def refresh(self) -> None:
        """Re-fetch the underlying :class:`Workflow` from the server."""
        data = self._ws._http.get_json(
            f"/v1/workspaces/{self._ws.id}/workflows/{self.id}",
            not_found_class=WorkflowNotFound,
        )
        self._wf = Workflow.model_validate(data)

    # -- v0.5: lease + worker integration ------------------------------

    def pending_steps(self) -> list[WorkflowStep]:
        """List steps in ``pending`` status (ready for a worker to lease).

        Workers use this to discover work; the v0.2 in-process flow
        creates steps in ``running`` directly so the list is empty
        unless a worker is in the loop.
        """

        data = self._ws._http.get_json(
            f"/v1/workspaces/{self._ws.id}/workflows/{self.id}/pending",
            not_found_class=WorkflowNotFound,
        )
        return [WorkflowStep.model_validate(s) for s in data.get("steps", [])]

    def expired_leases(self) -> list[Lease]:
        """Return leases past their expiry that haven't been reaped yet."""
        data = self._ws._http.get_json(
            f"/v1/workspaces/{self._ws.id}/workflows/{self.id}/expired",
            not_found_class=WorkflowNotFound,
        )
        return [Lease.model_validate(le) for le in data.get("leases", [])]

    def lease_step(
        self,
        step_id: str,
        worker_id: str,
        *,
        ttl: int = 60,
    ) -> Lease | None:
        """Try to acquire a lease on ``step_id`` for ``worker_id``.

        Returns the new :class:`Lease` on success or ``None`` on a 409
        ``LEASE_CONFLICT`` (someone else got it). All other errors
        propagate as the corresponding :class:`PlinthError` subclass.
        """

        from .exceptions import LeaseConflict

        try:
            response = self._ws._http.post(
                f"/v1/workspaces/{self._ws.id}/workflows/{self.id}"
                f"/steps/{step_id}/lease",
                json={"worker_id": worker_id, "ttl_seconds": ttl},
                not_found_class=WorkflowStepNotFound,
            )
        except LeaseConflict:
            return None
        return Lease.model_validate(response.json())

    def heartbeat_step(
        self,
        step_id: str,
        worker_id: str,
        *,
        ttl: int | None = None,
    ) -> Lease:
        """Extend the lease on ``step_id`` (must be held by ``worker_id``)."""
        body: dict[str, Any] = {"worker_id": worker_id}
        if ttl is not None:
            body["ttl_seconds"] = ttl
        response = self._ws._http.post(
            f"/v1/workspaces/{self._ws.id}/workflows/{self.id}"
            f"/steps/{step_id}/heartbeat",
            json=body,
            not_found_class=WorkflowStepNotFound,
        )
        return Lease.model_validate(response.json())

    def release_step(
        self,
        step_id: str,
        worker_id: str,
        *,
        status: str = "completed",
        output: Any = None,
        error: str | None = None,
        snapshot_id: str | None = None,
    ) -> Lease:
        """Release the lease, marking the step ``status`` (or pending to retry)."""
        body: dict[str, Any] = {"worker_id": worker_id, "status": status}
        if output is not None:
            body["output"] = output
        if error is not None:
            body["error"] = error
        if snapshot_id is not None:
            body["snapshot_id"] = snapshot_id
        response = self._ws._http.post(
            f"/v1/workspaces/{self._ws.id}/workflows/{self.id}"
            f"/steps/{step_id}/release",
            json=body,
            not_found_class=WorkflowStepNotFound,
        )
        lease = Lease.model_validate(response.json())
        # Refresh cached steps so callers can read the new status.
        self.refresh()
        return lease

    # -- v1.1: dead-letter queue ---------------------------------------

    def dlq(self) -> list[DLQEntry]:
        """List every DLQ entry recorded for this workflow.

        Entries are returned newest-first by ``failed_at``. An empty
        list is returned when no step has yet exhausted its retries.
        """
        data = self._ws._http.get_json(
            f"/v1/workspaces/{self._ws.id}/workflows/{self.id}/dlq",
            not_found_class=WorkflowNotFound,
        )
        return [DLQEntry.model_validate(e) for e in data.get("entries", [])]

    def replay_dlq(self, dlq_id: str) -> WorkflowStep:
        """Replay ``dlq_id`` as a fresh attempt of the same step name.

        The server creates a new step row in ``pending`` status (so a
        worker can immediately lease it) and deletes the DLQ entry in
        the same transaction. Returns the new :class:`WorkflowStep`.
        """
        response = self._ws._http.post(
            f"/v1/workspaces/{self._ws.id}/workflows/{self.id}"
            f"/dlq/{dlq_id}/replay",
            not_found_class=WorkflowNotFound,
        )
        body = response.json()
        # Refresh the cached workflow so the new step lands in the
        # handle's step log.
        self.refresh()
        return WorkflowStep.model_validate(body["replayed_step"])

    def delete_dlq(self, dlq_id: str) -> None:
        """Delete a DLQ entry without replaying it (operator dismissal)."""
        # No JSON response is expected (204), so we drop down to the
        # raw httpx client.
        response = self._ws._http._client.request(
            "DELETE",
            f"/v1/workspaces/{self._ws.id}/workflows/{self.id}/dlq/{dlq_id}",
        )
        self._ws._http._raise_for_status(
            response,
            not_found_class=WorkflowNotFound,
        )

    # -- internal ------------------------------------------------------

    def _record_step(self, step: WorkflowStep) -> None:
        """Update the cached steps list with the latest server view."""
        existing = self._wf.steps
        for i, s in enumerate(existing):
            if s.id == step.id:
                existing[i] = step
                return
        existing.append(step)


# ---------------------------------------------------------------------------
# WorkflowsProxy -- exposed via ``ws.workflows``.
# ---------------------------------------------------------------------------


class WorkflowsProxy:
    """API surface for the workspace's workflows.

    Held as ``ws.workflows`` (lazily instantiated on first attribute
    access). Returns :class:`WorkflowHandle` objects -- never the raw
    model -- so callers can chain step transitions in one expression.
    """

    def __init__(self, workspace: Workspace) -> None:
        self._ws = workspace

    # ------------------------------------------------------------------
    # create / get_or_create / get / list
    # ------------------------------------------------------------------

    def create(
        self,
        name: str,
        steps: list[str] | list[dict[str, Any]],
        metadata: dict[str, Any] | None = None,
    ) -> WorkflowHandle:
        """Create a new workflow with the given step manifest.

        v1.1: ``steps`` may now be a list of dicts carrying per-step
        retry configuration. The dict shape is::

            {"name": "search", "max_attempts": 3,
             "retry_policy": "exponential",
             "retry_initial_delay_seconds": 2.0,
             "retry_max_delay_seconds": 60.0,
             "retry_jitter": True}

        Only ``name`` is required. The retry config is cached on the
        handle so subsequent :meth:`WorkflowHandle.start_step` calls
        forward it to the server when the matching step is started.
        Bare-string entries are still accepted (v1.0 behaviour) — those
        steps default to ``max_attempts=1`` (no retry).

        Args:
            name: A human-readable name (e.g. ``"research-pipeline"``).
            steps: Ordered manifest. Either a list of step names, or a
                list of per-step config dicts (see above). Each
                ``start_step`` call must use a name from this list.
            metadata: Optional free-form metadata dict.

        Returns:
            A :class:`WorkflowHandle` wrapping the newly created workflow.
        """
        # Normalise: server expects ``list[str]`` for ``steps``. Per-step
        # retry config is cached on the handle so start_step() can pull
        # it back out without an extra round-trip.
        names: list[str] = []
        retry_cfg: dict[str, dict[str, Any]] = {}
        for entry in steps:
            if isinstance(entry, str):
                names.append(entry)
                continue
            if not isinstance(entry, dict) or "name" not in entry:
                raise ValueError(
                    "steps entries must be a string or a dict with a "
                    "'name' key"
                )
            names.append(entry["name"])
            cfg = {k: v for k, v in entry.items() if k != "name"}
            if cfg:
                retry_cfg[entry["name"]] = cfg

        body: dict[str, Any] = {"name": name, "steps": names}
        if metadata is not None:
            body["metadata"] = metadata

        response = self._ws._http.post(
            f"/v1/workspaces/{self._ws.id}/workflows",
            json=body,
            not_found_class=WorkspaceNotFound,
        )
        handle = WorkflowHandle(self._ws, Workflow.model_validate(response.json()))
        # Stash the retry config on the handle so start_step() can pick
        # it up by step name.
        handle._retry_config = retry_cfg  # type: ignore[attr-defined]
        return handle

    def get_or_create(
        self,
        name: str,
        steps: list[str],
        metadata: dict[str, Any] | None = None,
    ) -> WorkflowHandle:
        """Idempotent create-by-name: if a workflow with ``name`` already
        exists, fetch and return it; otherwise create one.

        The list lookup is best-effort -- if multiple workflows share the
        same name (legacy or test data), the first one returned by the
        server wins. New code should use unique names.
        """
        existing = next(
            (w for w in self.list() if w.name == name),
            None,
        )
        if existing is not None:
            # Re-fetch to load the full step log; ``list`` typically
            # returns a slimmer summary.
            return self.get(existing.id)
        return self.create(name, steps, metadata)

    def get(self, workflow_id: str) -> WorkflowHandle:
        """Fetch a workflow by ID (404 -> :class:`WorkflowNotFound`)."""
        data = self._ws._http.get_json(
            f"/v1/workspaces/{self._ws.id}/workflows/{workflow_id}",
            not_found_class=WorkflowNotFound,
        )
        return WorkflowHandle(self._ws, Workflow.model_validate(data))

    def list(self) -> list[Workflow]:
        """List every workflow on the workspace as :class:`Workflow` rows.

        Returns the bare model so the call cost is one request -- to act
        on a workflow, follow up with :meth:`get` for a
        :class:`WorkflowHandle`.
        """
        data = self._ws._http.get_json(
            f"/v1/workspaces/{self._ws.id}/workflows",
            not_found_class=WorkspaceNotFound,
        )
        return [Workflow.model_validate(w) for w in data.get("workflows", [])]


__all__ = ["WorkflowHandle", "WorkflowsProxy"]

File: sdk/go/plinth/workflows.go
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 The Plinth Authors

package plinth

import (
	"context"
	"errors"
)

// WorkflowsProxy is the v0.2 durable workflow surface for a workspace.
//
// Returns *WorkflowHandle from every non-list method so callers can
// chain step transitions in one expression.
type WorkflowsProxy struct {
	ws *WorkspaceClient
}

func newWorkflowsProxy(ws *WorkspaceClient) *WorkflowsProxy { return &WorkflowsProxy{ws: ws} }

// Create starts a new workflow with the given step manifest. metadata
// may be nil.
func (w *WorkflowsProxy) Create(ctx context.Context, name string, steps []string) (*WorkflowHandle, error) {
	return w.CreateWithMetadata(ctx, name, steps, nil)
}

// CreateWithMetadata is the variant of Create that accepts an
// arbitrary metadata bag persisted alongside the workflow.
func (w *WorkflowsProxy) CreateWithMetadata(ctx context.Context, name string, steps []string, metadata map[string]any) (*WorkflowHandle, error) {
	body := map[string]any{
		"name":  name,
		"steps": steps,
	}
	if metadata != nil {
		body["metadata"] = metadata
	}
	var wf Workflow
	err := w.ws.http.PostJSON(
		ctx,
		w.basePath(),
		&wf,
		WithJSON(body),
		WithNotFoundCode(ErrWorkspaceNotFound.Code),
	)
	if err != nil {
		return nil, err
	}
	return newWorkflowHandle(w.ws, wf), nil
}

// Get fetches a workflow by ID.
func (w *WorkflowsProxy) Get(ctx context.Context, workflowID string) (*WorkflowHandle, error) {
	var wf Workflow
	err := w.ws.http.GetJSON(
		ctx,
		w.basePath()+"/"+EncodePathSegment(workflowID),
		&wf,
		WithNotFoundCode(ErrWorkflowNotFound.Code),
	)
	if err != nil {
		return nil, err
	}
	return newWorkflowHandle(w.ws, wf), nil
}

// List returns every workflow on the workspace as bare models. Use
// Get to act on a workflow (it returns a *WorkflowHandle).
func (w *WorkflowsProxy) List(ctx context.Context) ([]Workflow, error) {
	var resp struct {
		Workflows []Workflow `json:"workflows"`
	}
	err := w.ws.http.GetJSON(
		ctx,
		w.basePath(),
		&resp,
		WithNotFoundCode(ErrWorkspaceNotFound.Code),
	)
	if err != nil {
		return nil, err
	}
	return resp.Workflows, nil
}

// GetOrCreate is an idempotent create-by-name. If a workflow with
// `name` already exists, returns it; otherwise creates one with the
// supplied manifest.
func (w *WorkflowsProxy) GetOrCreate(ctx context.Context, name string, steps []string) (*WorkflowHandle, error) {
	all, err := w.List(ctx)
	if err != nil {
		return nil, err
	}
	for _, wf := range all {
		if wf.Name == name {
			return w.Get(ctx, wf.ID)
		}
	}
	return w.Create(ctx, name, steps)
}

func (w *WorkflowsProxy) basePath() string {
	return "/v1/workspaces/" + EncodePathSegment(w.ws.ID()) + "/workflows"
}

// ---------------------------------------------------------------------------
// WorkflowHandle — methods on a single workflow
// ---------------------------------------------------------------------------

// WorkflowHandle wraps a *Workflow with method-style step transitions
// plus refresh/resume/lease helpers. Returned by every WorkflowsProxy
// method except List.
//
// Holds the parent workspace's HTTPClient by reference so callers
// don't have to thread anything through. Mutations
// (StartStep / CompleteStep / ...) refresh the cached model so
// subsequent reads against the handle are consistent.
type WorkflowHandle struct {
	ws *WorkspaceClient
	wf Workflow
}

func newWorkflowHandle(ws *WorkspaceClient, wf Workflow) *WorkflowHandle {
	return &WorkflowHandle{ws: ws, wf: wf}
}

// ID returns the workflow ID (e.g. "wf_01H…").
func (h *WorkflowHandle) ID() string { return h.wf.ID }

// Name returns the human-readable workflow name.
func (h *WorkflowHandle) Name() string { return h.wf.Name }

// Status returns the cached workflow status.
func (h *WorkflowHandle) Status() WorkflowStatus { return h.wf.Status }

// StepsManifest returns the expected step names in declaration order.
func (h *WorkflowHandle) StepsManifest() []string { return h.wf.StepsManifest }

// Steps returns the cached step log.
func (h *WorkflowHandle) Steps() []WorkflowStep { return h.wf.Steps }

// Model returns the cached *Workflow model. Use Refresh to re-fetch.
func (h *WorkflowHandle) Model() Workflow { return h.wf }

// StartStepOpts customises a single StartStep call.
type StartStepOpts struct {
	// Input is an optional payload recorded on the step.
	Input any
	// SnapshotID is an optional snapshot taken before the step ran.
	SnapshotID string
	// InitialStatus, when "pending", stages the step for a v0.5
	// durable worker to lease and run. Defaults to "running" for the
	// in-process flow.
	InitialStatus string
}

// StartStep records a new step on the workflow and returns it. Validates
// name against the manifest client-side so callers get a synchronous
// error before paying for the HTTP roundtrip.
func (h *WorkflowHandle) StartStep(ctx context.Context, name string, input any) (*WorkflowStep, error) {
	return h.StartStepWithOpts(ctx, name, StartStepOpts{Input: input})
}

// StartStepWithOpts is the variant of StartStep that exposes the full
// option struct (snapshot binding, initial status).
func (h *WorkflowHandle) StartStepWithOpts(ctx context.Context, name string, opts StartStepOpts) (*WorkflowStep, error) {
	if len(h.wf.StepsManifest) > 0 && !contains(h.wf.StepsManifest, name) {
		return nil, &PlinthError{
			Code:    ErrInvalidWorkflowStep.Code,
			Message: "step " + name + " is not declared in the workflow manifest",
		}
	}
	body := map[string]any{"name": name}
	if opts.Input != nil {
		body["input"] = opts.Input
	}
	if opts.SnapshotID != "" {
		body["snapshot_id"] = opts.SnapshotID
	}
	if opts.InitialStatus != "" {
		body["initial_status"] = opts.InitialStatus
	}
	var step WorkflowStep
	err := h.ws.http.PostJSON(
		ctx,
		h.stepsPath(),
		&step,
		WithJSON(body),
		WithNotFoundCode(ErrWorkflowNotFound.Code),
	)
	if err != nil {
		return nil, err
	}
	h.recordStep(step)
	return &step, nil
}

// CompleteStep marks stepID completed. snapshotID is the canonical
// resume point — pass "" to skip.
func (h *WorkflowHandle) CompleteStep(ctx context.Context, stepID string, output any, snapshotID string) (*WorkflowStep, error) {
	body := map[string]any{"status": "completed"}
	if output != nil {
		body["output"] = output
	}
	if snapshotID != "" {
		body["snapshot_id"] = snapshotID
	}
	return h.patchStep(ctx, stepID, body)
}

// FailStep marks stepID failed with a free-text error string.
func (h *WorkflowHandle) FailStep(ctx context.Context, stepID, errorMsg string) (*WorkflowStep, error) {
	return h.patchStep(ctx, stepID, map[string]any{
		"status": "failed",
		"error":  errorMsg,
	})
}

// CancelStep marks stepID cancelled.
func (h *WorkflowHandle) CancelStep(ctx context.Context, stepID string) (*WorkflowStep, error) {
	return h.patchStep(ctx, stepID, map[string]any{"status": "cancelled"})
}

// Cancel cancels the entire workflow on the server and refreshes the
// cached model.
func (h *WorkflowHandle) Cancel(ctx context.Context) error {
	var updated Workflow
	err := h.ws.http.PostJSON(
		ctx,
		h.workflowPath()+"/cancel",
		&updated,
		WithNotFoundCode(ErrWorkflowNotFound.Code),
	)
	if err != nil {
		return err
	}
	h.wf = updated
	return nil
}

// ResumeInfo returns the next pending step plus the snapshot to
// restore from. Crash → restart → call this → restore from
// SnapshotID → continue at NextStep.
func (h *WorkflowHandle) ResumeInfo(ctx context.Context) (*ResumeInfo, error) {
	var info ResumeInfo
	err := h.ws.http.GetJSON(
		ctx,
		h.workflowPath()+"/resume",
		&info,
		WithNotFoundCode(ErrWorkflowNotFound.Code),
	)
	if err != nil {
		return nil, err
	}
	return &info, nil
}

// Refresh re-fetches the workflow (with its full step log) from the
// server, replacing the cached model.
func (h *WorkflowHandle) Refresh(ctx context.Context) error {
	var wf Workflow
	err := h.ws.http.GetJSON(
		ctx,
		h.workflowPath(),
		&wf,
		WithNotFoundCode(ErrWorkflowNotFound.Code),
	)
	if err != nil {
		return err
	}
	h.wf = wf
	return nil
}

// PendingSteps returns steps in `pending` status — ready for a worker
// to lease.
func (h *WorkflowHandle) PendingSteps(ctx context.Context) ([]WorkflowStep, error) {
	var resp struct {
		Steps []WorkflowStep `json:"steps"`
	}
	err := h.ws.http.GetJSON(
		ctx,
		h.workflowPath()+"/pending",
		&resp,
		WithNotFoundCode(ErrWorkflowNotFound.Code),
	)
	if err != nil {
		return nil, err
	}
	return resp.Steps, nil
}

// ExpiredLeases returns leases past their expiry that haven't been
// reaped yet.
func (h *WorkflowHandle) ExpiredLeases(ctx context.Context) ([]Lease, error) {
	var resp struct {
		Leases []Lease `json:"leases"`
	}
	err := h.ws.http.GetJSON(
		ctx,
		h.workflowPath()+"/expired",
		&resp,
		WithNotFoundCode(ErrWorkflowNotFound.Code),
	)
	if err != nil {
		return nil, err
	}
	return resp.Leases, nil
}

// LeaseStep tries to lease stepID for workerID. Returns (nil, nil) on
// 409 LEASE_CONFLICT (someone else got it). Other errors surface
// normally.
func (h *WorkflowHandle) LeaseStep(ctx context.Context, stepID, workerID string, ttlSeconds int) (*Lease, error) {
	if ttlSeconds <= 0 {
		ttlSeconds = 60
	}
	var lease Lease
	err := h.ws.http.PostJSON(
		ctx,
		h.stepsPath()+"/"+EncodePathSegment(stepID)+"/lease",
		&lease,
		WithJSON(map[string]any{
			"worker_id":   workerID,
			"ttl_seconds": ttlSeconds,
		}),
		WithNotFoundCode(ErrWorkflowStepNotFound.Code),
	)
	if err != nil {
		if errors.Is(err, ErrLeaseConflict) {
			return nil, nil
		}
		return nil, err
	}
	return &lease, nil
}

// HeartbeatStep extends the lease on stepID. Must be called by the
// holding worker.
func (h *WorkflowHandle) HeartbeatStep(ctx context.Context, stepID, workerID string, ttlSeconds int) (*Lease, error) {
	body := map[string]any{"worker_id": workerID}
	if ttlSeconds > 0 {
		body["ttl_seconds"] = ttlSeconds
	}
	var lease Lease
	err := h.ws.http.PostJSON(
		ctx,
		h.stepsPath()+"/"+EncodePathSegment(stepID)+"/heartbeat",
		&lease,
		WithJSON(body),
		WithNotFoundCode(ErrWorkflowStepNotFound.Code),
	)
	if err != nil {
		return nil, err
	}
	return &lease, nil
}

// ReleaseStep releases the lease on stepID, marking the step `status`.
// status is typically "completed" or "failed".
func (h *WorkflowHandle) ReleaseStep(ctx context.Context, stepID, workerID string, status string, output any, errorMsg, snapshotID string) (*Lease, error) {
	body := map[string]any{
		"worker_id": workerID,
		"status":    status,
	}
	if output != nil {
		body["output"] = output
	}
	if errorMsg != "" {
		body["error"] = errorMsg
	}
	if snapshotID != "" {
		body["snapshot_id"] = snapshotID
	}
	var lease Lease
	err := h.ws.http.PostJSON(
		ctx,
		h.stepsPath()+"/"+EncodePathSegment(stepID)+"/release",
		&lease,
		WithJSON(body),
		WithNotFoundCode(ErrWorkflowStepNotFound.Code),
	)
	if err != nil {
		return nil, err
	}
	// Best-effort refresh so callers reading h.Steps after a release
	// see the new lifecycle. Failure is silent — a refresh hiccup
	// shouldn't mask a successful release.
	_ = h.Refresh(ctx)
	return &lease, nil
}

func (h *WorkflowHandle) patchStep(ctx context.Context, stepID string, body map[string]any) (*WorkflowStep, error) {
	var step WorkflowStep
	err := h.ws.http.PatchJSON(
		ctx,
		h.stepsPath()+"/"+EncodePathSegment(stepID),
		&step,
		WithJSON(body),
		WithNotFoundCode(ErrWorkflowStepNotFound.Code),
	)
	if err != nil {
		return nil, err
	}
	h.recordStep(step)
	return &step, nil
}

func (h *WorkflowHandle) recordStep(step WorkflowStep) {
	for i := range h.wf.Steps {
		if h.wf.Steps[i].ID == step.ID {
			h.wf.Steps[i] = step
			return
		}
	}
	h.wf.Steps = append(h.wf.Steps, step)
}

func (h *WorkflowHandle) workflowPath() string {
	return "/v1/workspaces/" + EncodePathSegment(h.ws.ID()) + "/workflows/" + EncodePathSegment(h.wf.ID)
}

func (h *WorkflowHandle) stepsPath() string {
	return h.workflowPath() + "/steps"
}

// contains reports whether haystack contains needle. Tiny helper to
// avoid pulling in golang.org/x/exp/slices for a single check.
func contains(haystack []string, needle string) bool {
	for _, h := range haystack {
		if h == needle {
			return true
		}
	}
	return false
}
